src/adapter/src/optimize/mod.rs

```diff
@@ -0,0 +1,160 @@
+/// (`Index`, `View`, `MaterializedView`, `Subscribe`, `Select`).
+///
+/// Each implementation represents a concrete optimization stage for a fixed
+/// statement type that consumes an input of type `From` and produces output of
+/// type `Self::To`.
+///
+/// The generic lifetime `'ctx` models the lifetime of the optimizer context and
+/// can be passed to the optimizer struct and the `Self::To` types.
+///
+/// The `'s: 'ctx` bound in the `optimize` method call ensures that an optimizer
+/// instance can run an optimization stage that produces a `Self::To` with
+/// `&'ctx` references.
+pub trait Optimize<'ctx, From>: Send + Sync
+where
+    From: Send + Sync,
+{
+    type To: Send + Sync + 'ctx;
+
+    /// Execute the optimization stage, transforming the input plan of type
+    /// `From` to an output plan of type `To`.
+    fn optimize<'s: 'ctx>(&'s mut self, plan: From) -> Result<Self::To, OptimizerError>;
+
+    /// Execute the optimization stage and panic if an error occurs.
+    ///
+    /// See [`Optimize::optimize`].
+    fn must_optimize<'s: 'ctx>(&'s mut self, expr: From) -> Self::To {
+        match self.optimize(expr) {
+            Ok(ok) => ok,
+            Err(err) => panic!("must_optimize call failed: {err}"),
+        }
+    }
+}
+
+// Feature flags for the optimizer.
+pub struct OptimizerFlags {
+    /// Eanble consolidation of unions that happen immediately after negate.
+    ///
+    /// The refinement happens in the LIR ⇒ LIR phase.
+    pub enable_consolidate_after_union_negate: bool,
+    /// Eanble relaxing of consolidation enforcers for optimization pipelines
+    /// that produce dataflows that run a computation for a single timestamp
+    /// (such as SELECTs).
+    ///
+    /// The refinement happens in the LIR ⇒ LIR phase.
+    pub enable_monotonic_oneshot_selects: bool,
+    /// An exclusive upper bound on the number of results we may return from a
+    /// Persist fast-path peek.
+    pub persist_fast_path_limit: usize,
+}
+
+impl From<&SystemVars> for OptimizerFlags {
+    fn from(vars: &SystemVars) -> Self {
+        Self {
+            enable_consolidate_after_union_negate: vars.enable_consolidate_after_union_negate(),
+            enable_monotonic_oneshot_selects: vars.enable_monotonic_oneshot_selects(),
+            persist_fast_path_limit: vars.persist_fast_path_limit(),
+        }
+    }
+}
+
+/// A type for a [`DataflowDescription`] backed by `Mir~` plans. Used internally
+/// by the optimizer implemenations.
+type MirDataflowDescription = DataflowDescription<OptimizedMirRelationExpr>;
+/// A type for a [`DataflowDescription`] backed by `Lir~` plans. Used internally
+/// by the optimizer implemenations.
+type LirDataflowDescription = DataflowDescription<Plan>;
+
+/// Error types that can be generated during optimization.
+#[derive(Debug, thiserror::Error)]
+pub enum OptimizerError {
+    // TODO: change dataflows.rs error types and reverse this ownership.
```

Comment:
Curious how this would work! Note that we cannot have backwards-incompatible changes for dataflow errors, but can't tell from the comment that's related.

Reply:
I don't think this is related. We need to change the method signatures in `adapter/src/coord/dataflows.rs` to return `OptimizerError` instead of `AdapterError`.